### DISCUSSION TITLE
Progress bar doesn't update

### DISCUSSION THREAD

**Discussion by TangDave** [AvaloniaUI/Avalonia#15953] [Q&A] [answered]
I have a window that shows when another process is completing. The other process sets a boolean flag when done - a function returning the status of this flag is passed to the window (this is the function referred to by _signalCompletedFunction). The window 'winWait' just shows a cycling progress bar (moves from 1 to 10 repeatedly while waiting). The problem is that the progress bar never updates. (Similar code works in WinForms where 'OnLoaded' is replaced by WinForms 'OnActivated', but the WinForms progress bar 'Refresh' method is also required - this method doesn't exist in Avalonia).

```
    public partial class winWait : Window
    {
    	private readonly Func<bool> _signalCompletedFunction;
    	
    	public winWait()
    	{
    		InitializeComponent();
    	}
    
    	public winWait(Func<bool> signalCompletedFunction) : this()
    	{
    		_signalCompletedFunction = signalCompletedFunction;
    
    		this.prgWait.Minimum = 0;
    		this.prgWait.Maximum = 10;
    	}
    
    	protected override void OnLoaded(RoutedEventArgs e)
    	{
    		base.OnLoaded(e);
    
    		while (!_signalCompletedFunction())
    		{
    			double mod = (this.prgWait.Value + 1) % this.prgWait.Maximum;
    			this.prgWait.Value = mod == 0 ? this.prgWait.Maximum : mod;
    		}
    
    		this.Close();
    	}
    }
```

**Comment by stevemonaco** [comment, marked as answer]
This looks like you're blocking the UI thread with a long-running loop that also runs on the UI thread. Your entire app should be frozen and the behavior should be the same on WinForms if the implementation there is similar.

Maybe change the event handler to `async void` and put an `await Task.Delay(100);` or such inside of the loop? You don't need to be checking that state and setting the `ProgressBar` many thousands of times / sec or more.

**Reply by TangDave** [reply]
Thanks! That solved it nicely.

In WinForms, I used a Thread.Sleep(200) combined with a Refresh call which made it work there - the Thread.Sleep did not work at all with the Avalonia version, but making the OnLoaded method async with the Task.Delay worked great.

**Reply by stevemonaco** [reply]
It must have been the refresh call that made it work. `Thread.Sleep` blocks the current thread entirely to do nothing and if that's your UI thread...then it doesn't render/update.
